File: r/svr.R
# Support Vector Regression

# Import dataset
dataset <- read.csv('../data/Position_Salaries.csv')
dataset <- dataset[, 2:3]

# Fit the SVR model
library(e1071)
regressor <- svm(Salary ~ ., data = dataset, type = 'eps-regression')

# Predict result for level 6.5
y_pred <- predict(regressor, newdata = data.frame(Level = 6.5))
cat("💰 Predicted salary for level 6.5:", y_pred, "\n")

# Base scatter plot with original data
library(ggplot2)

# Visualizing the SVR results (original resolution)
ggplot() +
  geom_point(data = dataset, aes(x = Level, y = Salary), color = 'red', size = 2) +
  geom_line(data = dataset, aes(x = Level, y = predict(regressor, newdata = dataset)), color = 'blue', size = 1) +
  labs(
    title = "Salary vs Level (Support Vector Regression)",
    x = "Level",
    y = "Salary"
  ) +
  theme_minimal()

# === High-resolution version ===
# Create a finer grid for smooth plotting
X_grid <- seq(min(dataset$Level), max(dataset$Level), by = 0.01)
y_pred_grid <- predict(regressor, newdata = data.frame(Level = X_grid))

# Plot the high-resolution curve
ggplot() +
  geom_point(data = dataset, aes(x = Level, y = Salary), color = 'red', size = 2) +
  geom_line(aes(x = X_grid, y = y_pred_grid), color = 'blue', size = 1) +
  labs(
    title = "Salary vs Level (SVR - High Resolution)",
    x = "Level",
    y = "Salary"
  ) +
  theme_minimal()

File: python/svr.py
# Support Vector Regression (SVR)

# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR

# Load Dataset
dataset = pd.read_csv('../data/Position_Salaries.csv')
dataset = dataset.iloc[:, 1:]  # Remove the first column (Position)

# Extract Features and Target
X = dataset.iloc[:, [0]].values  # Level
y = dataset.iloc[:, -1].values   # Salary

# Feature Scaling
sc_X = StandardScaler()
sc_y = StandardScaler()

X_scaled = sc_X.fit_transform(X)
y_scaled = sc_y.fit_transform(y.reshape(-1, 1)).ravel()  # Flatten to 1D

# Fit the SVR Model
regressor = SVR(kernel='rbf')
regressor.fit(X_scaled, y_scaled)

# Predict a new result (e.g., for Level 6.5)
level_input = [[6.5]]
scaled_input = sc_X.transform(level_input)
scaled_prediction = regressor.predict(scaled_input)
prediction = sc_y.inverse_transform(scaled_prediction.reshape(-1, 1))

print(f"💰 Predicted Salary for Level 6.5: ${prediction[0, 0]:,.2f}")

# Visualize SVR Results
plt.figure(figsize=(8, 5))
plt.scatter(X_scaled, y_scaled, color='red', label='Original Data')
plt.plot(X_scaled, regressor.predict(X_scaled), color='blue', label='SVR Model')
plt.title('Support Vector Regression (SVR)')
plt.xlabel('Level (scaled)')
plt.ylabel('Salary (scaled)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

# Visualize SVR in High Resolution
X_grid = np.linspace(X_scaled.min(), X_scaled.max(), 100).reshape(-1, 1)

plt.figure(figsize=(8, 5))
plt.scatter(X_scaled, y_scaled, color='red', label='Original Data')
plt.plot(X_grid, regressor.predict(X_grid), color='blue', label='SVR High Res')
plt.title('Support Vector Regression (High Resolution)')
plt.xlabel('Level (scaled)')
plt.ylabel('Salary (scaled)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()
